File: c/sorting/standard/intro_sort.c
#include <stdlib.h>
#include <math.h>

int* heap_sort(int *arr, int n);
int* insertion_sort(int *arr, int n);

int* _intro_sort(int *a, int n, int max_depth)
{
    if (n <= 16)
        return insertion_sort(a, n);
    if (max_depth == 0)
        return heap_sort(a, n);

    int pivot = a[n / 2];
    int *left = malloc(sizeof(int) * n);
    int *middle = malloc(sizeof(int) * n);
    int *right = malloc(sizeof(int) * n);
    int l = 0, m = 0, r = 0;

    for (int i = 0; i < n; i++)
    {
        if (a[i] < pivot) left[l++] = a[i];
        else if (a[i] == pivot) middle[m++] = a[i];
        else right[r++] = a[i];
    }

    int *sorted_left = _intro_sort(left, l, max_depth - 1);
    int *sorted_right = _intro_sort(right, r, max_depth - 1);
    int *result = malloc(sizeof(int) * n);
    int pos = 0;

    for (int i = 0; i < l; i++) result[pos++] = sorted_left[i];
    for (int i = 0; i < m; i++) result[pos++] = middle[i];
    for (int i = 0; i < r; i++) result[pos++] = sorted_right[i];

    return result;
}

int* intro_sort(int *arr, int n)
{
    int max_depth = 2 * ((int)log2(n) + 1);
    return _intro_sort(arr, n, max_depth);
}

File: python/sorting/standard/intro_sort.py
from .heap_sort import heap_sort
from .insertion_sort import insertion_sort


def intro_sort(arr):
    max_depth = 2 * len(arr).bit_length()
    return _intro_sort(arr[:], max_depth)


def _intro_sort(a, max_depth):
    if len(a) <= 16:
        return insertion_sort(a)

    if max_depth == 0:
        return heap_sort(a)

    pivot = a[len(a) // 2]
    left = [x for x in a if x < pivot]
    middle = [x for x in a if x == pivot]
    right = [x for x in a if x > pivot]

    return _intro_sort(left, max_depth - 1) + middle + _intro_sort(right, max_depth - 1)
